### PR TITLE
Enable errorprone build

### PR DESCRIPTION
**Pull request by moltmann**
- fix up issues found
- placate error prone warning where we know we do something funky
- Ignore two tests as it is not immediately clear what is broken
  - org.mockitousage.bugs.varargs.VarargsNotPlayingWithAnyObjectTest#shouldNotAllowUsingAnyObjectForVarArgs
  - org.mockito.internal.util.PrimitivesTest#should_check_that_value_type_is_assignable_to_wrapper_reference

NOTE: Please carefully review the changes made to the code. I fixed them up as good as I could but a real mockito expert need to verify them.

### CODE REVIEW

File: src/main/java/org/mockito/internal/util/Primitives.java [hunk 1]
```diff
@@ -44,7 +44,7 @@ public static boolean isPrimitiveOrWrapper(Class<?> type) {
 
     public static boolean isAssignableFromWrapper(Class<?> valueClass, Class<?> referenceType) {
         if(isPrimitiveOrWrapper(valueClass) && isPrimitiveOrWrapper(referenceType)) {
-            return Primitives.primitiveTypeOf(valueClass).isAssignableFrom(referenceType);
+            return Primitives.primitiveTypeOf(valueClass).isAssignableFrom(Primitives.primitiveTypeOf(referenceType));
```
Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/util/Primitives.java]: This seemed to be the most sensible fix. It did not break any other test :thinking:

Comment (mockitoguy) [reply src/main/java/org/mockito/internal/util/Primitives.java]: If it does not break tests, we're good ;) This strategy applies to 99% of our codebase. We have few weak spots that are not covered with tests (like OSGi bundle configuration), but primitives handling should be good.

Thanks for checking!

File: src/test/java/org/mockito/internal/util/PrimitivesTest.java [hunk 2]
```diff
@@ -69,21 +69,21 @@ public void should_return_null_for_everything_else() throws Exception {
...
-        assertThat(Primitives.isAssignableFromWrapper(Float.class, Float.class));
-        assertThat(Primitives.isAssignableFromWrapper(char.class, Character.class));
-        assertThat(Primitives.isAssignableFromWrapper(Character.class, Character.class));
-        assertThat(Primitives.isAssignableFromWrapper(short.class, Short.class));
-        assertThat(Primitives.isAssignableFromWrapper(Short.class, Short.class));
-        assertThat(Primitives.isAssignableFromWrapper(byte.class, Byte.class));
-        assertThat(Primitives.isAssignableFromWrapper(Byte.class, Byte.class));
-        assertThat(Primitives.isAssignableFromWrapper(boolean.class, Boolean.class));
-        assertThat(Primitives.isAssignableFromWrapper(Boolean.class, Boolean.class));
+        assertThat(Primitives.isAssignableFromWrapper(int.class, Integer.class)).isTrue();
```
Comment (mockitoguy) [inline src/test/java/org/mockito/internal/util/PrimitivesTest.java]: So previously we didn't have any assertions here, right? Nice :)

GREAT CATCH, thanks!

File: src/test/java/org/mockitousage/bugs/varargs/VarargsNotPlayingWithAnyObjectTest.java [hunk 3]
```diff
@@ -35,18 +37,6 @@ public void shouldMatchAnyVararg() {
         verify(mock, never()).run(anyString(), eq("f"));
     }
 
-    //we cannot use anyObject() for entire varargs because it makes the verification pick up extra invocations
```
Comment (mockitoguy) [inline src/test/java/org/mockitousage/bugs/varargs/VarargsNotPlayingWithAnyObjectTest.java]: Are you removing the test because errorprone does not like empty catch statement? Hmm, it is not a good (even though I love removing code :). If you propose to remove it, please validate carefully and ensure we're not reducing coverage.

From my initial look, the test is valid, even though it covers a corner case. I suggest to either suppress the warning or use AssertJ's assertThrownBy

File: src/test/java/org/mockitousage/matchers/MoreMatchersTest.java [hunk 4]
```diff
@@ -48,12 +48,14 @@ public void any_class_should_be_actual_alias_to_isA() {
         mock.simpleMethod((String) null);
         try {
             verify(mock).simpleMethod(isA(String.class));
-            fail();
-        } catch (AssertionError ignored) { }
-        try {
-            verify(mock).simpleMethod(any(String.class));
-            fail();
-        } catch (AssertionError ignored) { }
+        } catch (AssertionError ignored) {
```
Comment (mockitoguy) [inline src/test/java/org/mockitousage/matchers/MoreMatchersTest.java]: The resulting code is less readable. I suggest AssertJ or suppression.

File: src/test/java/org/mockitousage/verification/BasicVerificationTest.java [hunk 5]
```diff
@@ -47,8 +47,11 @@ public void shouldFailVerificationOnMethodArgument() throws Exception {
         verify(mock).clear();
         try {
             verify(mock).add("bar");
-            fail();
-        } catch (AssertionError expected) {}
+        } catch (AssertionError expected) {
```
Comment (mockitoguy) [inline src/test/java/org/mockitousage/verification/BasicVerificationTest.java]: Empty catch again? :) AssertJ or suppress will work better for us I think.

File: src/test/java/org/mockitousage/bugs/varargs/VarargsNotPlayingWithAnyObjectTest.java (hunk 3 above)
Comment (TimvdLippe) [reply src/test/java/org/mockitousage/bugs/varargs/VarargsNotPlayingWithAnyObjectTest.java]: It was because we were actually not asserting anything in this case. The reason is that the `fail()` statement was always executed, because `verify(mock).run((String[]) anyObject());` is completely valid and works. This would then throw an `AssertionError`, which is consequently caught in the `try-catch` in the test case. In other words, whether the `fail` is reached or not, we would always catch the error and the test would never fail (even if there is a bug). Errorprone is correct in identifying that this is an issue.

File: src/test/java/org/mockitousage/matchers/MoreMatchersTest.java (hunk 4 above)
Comment (moltmann) [reply src/test/java/org/mockitousage/matchers/MoreMatchersTest.java]: done. Will look nicer once we can use lambdas

File: src/test/java/org/mockitousage/verification/BasicVerificationTest.java (hunk 5 above)
Comment (moltmann) [reply src/test/java/org/mockitousage/verification/BasicVerificationTest.java]: done. Will look nicer once we can use lambdas

File: src/test/java/org/mockitousage/bugs/varargs/VarargsNotPlayingWithAnyObjectTest.java (hunk 3 above)
Comment (TimvdLippe) [reply src/test/java/org/mockitousage/bugs/varargs/VarargsNotPlayingWithAnyObjectTest.java]: @mockitoguy Any update regarding this comment?

Comment (mockitoguy) [reply src/test/java/org/mockitousage/bugs/varargs/VarargsNotPlayingWithAnyObjectTest.java]: Oh I see! Thank you for clarification!

In that case we should absolutely clean up the test method. I'll push a commit with a desired fix and merge the PR.

Thank you for digging deep into this one. I missed the flaw in this test. Clean tests for the win!!!

**Comment by mockitoguy** [comment]
We're ready. I changed and kept the test that you identified as flawed (thanks for that!).

When Travis is green I'll merge.
